src/hotspot/share/gc/z/c2/zBarrierSetC2.cpp

```diff
@@ -300,13 +300,16 @@ void ZBarrierSetC2::clone_at_expansion(PhaseMacroExpand* phase, ArrayCopyNode* a
       // BarrierSetC2::clone sets the offsets via BarrierSetC2::arraycopy_payload_base_offset
       // which 8-byte aligns them to allow for word size copies. Make sure the offsets point
       // to the first element in the array when cloning object arrays. Otherwise, load
-      // barriers are applied to parts of the header.
+      // barriers are applied to parts of the header. Also adjust the length accordingly.
       assert(src_offset == dest_offset, "should be equal");
-      assert((src_offset->get_long() == arrayOopDesc::base_offset_in_bytes(T_OBJECT) && UseCompressedClassPointers) ||
-             (src_offset->get_long() == arrayOopDesc::length_offset_in_bytes() && !UseCompressedClassPointers),
-             "unexpected offset for object array clone");
-      src_offset = phase->longcon(arrayOopDesc::base_offset_in_bytes(T_OBJECT));
-      dest_offset = src_offset;
+      jlong offset = src_offset->get_long();
+      if (offset != arrayOopDesc::base_offset_in_bytes(T_OBJECT)) {
+        assert(!UseCompressedClassPointers, "should only happen without compressed class pointers");
+        assert((arrayOopDesc::base_offset_in_bytes(T_OBJECT) - offset) == BytesPerLong, "unexpected offset");
+        length = phase->transform_later(new SubXNode(length, phase->longcon(1))); // Size is in longs
```

Comment:
Do we have check that `length` is count of `BytesPerLong`?
Do we have a guard that this code is executed only in 64-bit VM?
`SubXNode` is SubINode in 32-bit VM.

Reply:
Z is 64 bit only.

Reply:
Okay. `SubLNode` should be used then to avoid confusion.
By expending code I see that it is guarded by `(bt == T_OBJECT)` so `length` is indeed is in `longs`. No issue here.

Reply:
Thanks for the review, Vladimir. Yes, ZGC is 64-bit only and the code I'm fixing is applied to object arrays only. I've changed `SubXNode` to `SubLNode` to avoid confusion.